apps/server/src/modules/tool/mapper/external-tool-request.mapper.ts

```diff
@@ -0,0 +1,118 @@
+	[CustomParameterTypeParams.AUTO_COURSENAME]: CustomParameterType.AUTO_COURSENAME,
+	[CustomParameterTypeParams.AUTO_SCHOOLID]: CustomParameterType.AUTO_SCHOOLID,
+};
+
+@Injectable()
+export class ExternalToolRequestMapper {
+	mapRequestToExternalToolDO(externalToolParams: ExternalToolParams, version: number): ExternalToolDO {
+		let mappedConfig: BasicToolConfigDO | Lti11ToolConfigDO | Oauth2ToolConfigDO;
+		if (externalToolParams.config instanceof BasicToolConfigParams) {
+			mappedConfig = this.mapBasicToolConfig(externalToolParams.config);
+		} else if (externalToolParams.config instanceof Lti11ToolConfigParams) {
+			mappedConfig = this.mapLti11ToolConfig(externalToolParams.config);
+		} else {
+			mappedConfig = this.mapOauth2ToolConfig(externalToolParams.config);
+		}
+
+		const mappedCustomParameter: CustomParameterDO[] = this.mapCustomParameterCreateParamsToDO(
+			externalToolParams.parameters ?? []
+		);
+
+		return new ExternalToolDO({
+			name: externalToolParams.name,
+			url: externalToolParams.url,
+			logoUrl: externalToolParams.logoUrl,
+			config: mappedConfig,
+			parameters: mappedCustomParameter,
+			isHidden: externalToolParams.isHidden,
+			openNewTab: externalToolParams.openNewTab,
+			version,
+		});
+	}
+
+	private mapBasicToolConfig(externalToolConfigParams: BasicToolConfigParams): BasicToolConfigDO {
+		return new BasicToolConfigDO({ ...externalToolConfigParams });
+	}
+
+	private mapLti11ToolConfig(externalToolConfigParams: Lti11ToolConfigParams): Lti11ToolConfigDO {
+		return new Lti11ToolConfigDO({ ...externalToolConfigParams });
+	}
+
+	private mapOauth2ToolConfig(externalToolConfigParams: Oauth2ToolConfigParams): Oauth2ToolConfigDO {
+		return new Oauth2ToolConfigDO({ ...externalToolConfigParams });
+	}
+
+	private mapCustomParameterCreateParamsToDO(
+		customParameterParams: CustomParameterCreateParams[]
+	): CustomParameterDO[] {
+		return customParameterParams.map((customParameterParam: CustomParameterCreateParams) => {
+			return new CustomParameterDO({
+				name: customParameterParam.name,
+				default: customParameterParam.default,
+				regex: customParameterParam.regex,
+				scope: scopeMapping[customParameterParam.scope],
+				location: locationMapping[customParameterParam.location],
+				type: typeMapping[customParameterParam.type],
+			});
+		});
+	}
+
+	mapDoToProviderOauthClient(name: string, oauth2Config: Oauth2ToolConfigDO): ProviderOauthClient {
+		return {
+			client_name: name,
+			client_id: oauth2Config.clientId,
+			client_secret: oauth2Config.clientSecret,
+			scope: oauth2Config.scope,
+			token_endpoint_auth_method: oauth2Config.tokenEndpointAuthMethod,
+			redirect_uris: oauth2Config.redirectUris,
+			frontchannel_logout_uri: oauth2Config.frontchannelLogoutUri,
+			subject_type: 'pairwise',
+		};
+	}
+
+	applyProviderOauthClientToDO(oauth2Config: Oauth2ToolConfigDO, oauthClient: ProviderOauthClient): Oauth2ToolConfigDO {
```

Comment:
I was wondering about the name in a mapper. Isnt it also an mapping? Why you named it apply..? I would expect a method called applyXYZ in the UC or in a Service

Reply:
because we extend the externalDO, it's not just a mapping..